File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/browser/State.java
/**
 * @file
 * @brief File browser state helper
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.ui.browser;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.Nullable;

import java.util.Locale;

import app.zxtune.preferences.DataStore;
import app.zxtune.preferences.Preferences;

class State {

  private static final String PREF_BROWSER = "browser_";
  private static final String PREF_BROWSER_CURRENT = PREF_BROWSER + "current";
  private static final String PREF_BROWSER_PATH_TEMPLATE = PREF_BROWSER + "%d_path";
  private static final String PREF_BROWSER_VIEWPOS_TEMPLATE = PREF_BROWSER + "%d_viewpos";
  private final DataStore prefs;
  private PathAndPosition current;

  private State(Context ctx) {
    this.prefs = Preferences.getDataStore(ctx);
    this.current = new PathAndPosition();
  }

  static State create(Context ctx) {
    return new State(ctx);
  }

  final Uri getCurrentPath() {
    return current.getPath();
  }

  final void setCurrentPath(Uri uri) {
    final Uri curPath = current.getPath();

    if (curPath.equals(uri)) {
      return;
    } else if (isNested(curPath, uri)) {
      setInnerPath(uri);
    } else if (isNested(uri, curPath)) {
      setOuterPath(uri);
    } else {
      setPath(uri);
    }
    current.store();
  }

  final int getCurrentViewPosition() {
    return current.getViewPosition();
  }

  final void setCurrentViewPosition(int pos) {
    current.setViewPosition(pos);
  }

  private void setInnerPath(Uri newPath) {
    current = new PathAndPosition(current.getIndex() + 1, newPath);
  }

  private void setOuterPath(Uri newPath) {
    current = findByPath(newPath);
  }

  private void setPath(Uri newPath) {
    current = new PathAndPosition(0, newPath);
  }

  private PathAndPosition findByPath(Uri newPath) {
    for (int idx = current.getIndex() - 1; idx >= 0; --idx) {
      final PathAndPosition pos = new PathAndPosition(idx);
      if (newPath.equals(pos.getPath())) {
        return pos;
      }
    }
    return new PathAndPosition(0, newPath);
  }

  // checks if rh is nested path relative to lh
  private static boolean isNested(Uri lh, Uri rh) {
    final String lhScheme = lh.getScheme();
    if (lhScheme != null && lhScheme.equals(rh.getScheme())) {
      final String rhPath = rh.getPath();
      final String lhPath = lh.getPath();
      if (rhPath != null) {
        if (lhPath == null) {
          return true;
        } else if (lhPath.equals(rhPath)) {
          return isNestedSubpath(lh.getFragment(), rh.getFragment());
        } else {
          return rhPath.startsWith(lhPath);
        }
      }
    }
    return false;
  }

  private static boolean isNestedSubpath(@Nullable String lh, @Nullable String rh) {
    return rh != null && (lh == null || rh.startsWith(lh));
  }

  private class PathAndPosition {

    private final int index;
    private final Uri path;
    private int position;

    PathAndPosition() {
      this(prefs.getInt(PREF_BROWSER_CURRENT, 0));
    }

    PathAndPosition(int idx) {
      this.index = idx;
      this.path = Uri.parse(prefs.getString(getPathKey(), ""));
      this.position = prefs.getInt(getPosKey(), 0);
    }

    PathAndPosition(int idx, Uri path) {
      this.index = idx;
      this.path = path;
      this.position = 0;
    }

    final int getIndex() {
      return index;
    }

    final Uri getPath() {
      return path;
    }

    final int getViewPosition() {
      return position;
    }

    final void setViewPosition(int newPos) {
      if (newPos != position) {
        position = newPos;
        store();
      }
    }

    private void store() {
      Bundle batch = new Bundle();
      batch.putString(getPathKey(), this.path.toString());
      batch.putInt(getPosKey(), this.position);
      batch.putInt(PREF_BROWSER_CURRENT, index);
      prefs.putBatch(batch);
    }

    private String getPathKey() {
      return String.format(Locale.US, PREF_BROWSER_PATH_TEMPLATE, index);
    }

    private String getPosKey() {
      return String.format(Locale.US, PREF_BROWSER_VIEWPOS_TEMPLATE, index);
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/browser/State.kt
/**
 * @file
 * @brief File browser state helper
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.ui.browser

import android.content.Context
import android.net.Uri
import android.os.Bundle
import androidx.annotation.VisibleForTesting
import app.zxtune.preferences.DataStore
import app.zxtune.preferences.Preferences.getDataStore
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

internal class State @VisibleForTesting constructor(
    ctx: Context, private val ioDispatcher: CoroutineDispatcher
) {
    private val prefs = getDataStore(ctx)
    private var _current: PathAndPosition? = null
    private var current
        get() = _current ?: PathAndPosition(prefs).also {
            _current = it
        }
        set(value) {
            _current = value
        }

    suspend fun getCurrentPath(): Uri = withContext(ioDispatcher) {
        current.path
    }

    suspend fun updateCurrentPath(uri: Uri) = withContext(ioDispatcher) {
        if (current.path != uri) {
            current = switchTo(uri).apply {
                save(this)
            }
        }
        current.viewPosition
    }

    private fun switchTo(uri: Uri) = when {
        isNested(current.path, uri) -> innerPath(uri)
        isNested(uri, current.path) -> outerPath(uri)
        else -> newPath(uri)
    }

    suspend fun updateCurrentPosition(pos: Int) = withContext(ioDispatcher) {
        current.run {
            viewPosition = pos
            save(this)
        }
    }

    private fun innerPath(newPath: Uri) = PathAndPosition(current.index + 1, newPath)

    private fun outerPath(newPath: Uri): PathAndPosition {
        for (idx in current.index - 1 downTo 0) {
            val pos = PathAndPosition(idx, prefs)
            if (newPath == pos.path) {
                return pos
            }
        }
        return PathAndPosition(0, newPath)
    }

    private fun newPath(newPath: Uri) = PathAndPosition(0, newPath)

    private fun save(data: PathAndPosition) {
        Bundle().apply {
            data.storeTo(this)
        }.also {
            prefs.putBatch(it)
        }
    }

    private class PathAndPosition {
        val index: Int
        val path: Uri
        var viewPosition: Int

        constructor(prefs: DataStore) : this(prefs.getInt(CURRENT_KEY, 0), prefs)

        constructor(idx: Int, prefs: DataStore) {
            index = idx
            path = Uri.parse(prefs.getString(pathKey, ""))
            viewPosition = prefs.getInt(posKey, 0)
        }

        constructor(idx: Int, path: Uri) {
            this.index = idx
            this.path = path
            this.viewPosition = 0
        }

        fun storeTo(bundle: Bundle) = with(bundle) {
            putString(pathKey, path.toString())
            putInt(posKey, viewPosition)
            putInt(CURRENT_KEY, index)
        }

        private val pathKey
            get() = "browser_${index}_path"

        private val posKey: String
            get() = "browser_${index}_viewpos"
    }

    companion object {
        private const val CURRENT_KEY = "browser_current"

        var create: (Context) -> State = { ctx -> State(ctx, Dispatchers.IO) }
            @VisibleForTesting set

        // checks if rh is nested path relative to lh
        private fun isNested(lh: Uri, rh: Uri) = lh.scheme == rh.scheme && rh.path?.let { rhPath ->
            when (val lhPath = lh.path) {
                null -> true
                rhPath -> isNestedSubpath(lh.fragment, rh.fragment)
                else -> rhPath.startsWith(lhPath)
            }
        } ?: false

        private fun isNestedSubpath(lh: String?, rh: String?) =
            rh != null && (lh == null || rh.startsWith(lh))
    }
}
